contributors/design-proposals/deployment.md

```diff
@@ -159,56 +165,89 @@ https://github.com/kubernetes/kubernetes/issues/11234 tracks updating kubelet
 and https://github.com/kubernetes/kubernetes/issues/12615 tracks adding
 LastTransitionTime to PodCondition.
 
+### TemplateGeneration
+
+Hashing an API object such as PodTemplateSpec means that the resulting hash is subject to
+change due to API changes in PodTemplateSpec between minor versions of Kubernetes when new
+API changes are introduced. A new API field will be introduced in DeploymentSpec called
+TemplateGeneration. The new field will be a counter that will track the number of
+times an update happened in the PodTemplateSpec of the Deployment, similarly to how
+metadata.Generation tracks updates in the Spec of all first-class API objects. Unlike
+metadata.Generation, this field can be initialized by users in order to facilitate adoption
+of existing ReplicaSets. This is similar to how it is already used by DaemonSets.
+
+TemplateGeneration is not authoritative and only helps in constructing the name for the new
+ReplicaSet in case there is no other ReplicaSet that matches the Deployment. The Deployment
+controller still decides what is the new ReplicaSet by comparing PodTemplateSpecs.
+If no matching ReplicaSet is found, the controller will try to create a new ReplicaSet using
+its current TemplateGeneration.
+
+The naming scheme used by ReplicaSets (*deployment.name-podtemplatehash*) will need to change
+to something different because we cannot migrate old ReplicaSets to use TemplateGeneration
+for their names and we want to avoid name collisions. For new RS names, we can either append
+the TemplateGeneration to the Deployment name or we can compute the hash of
+*deployment.name+deployment.templategeneration*, extend its size deterministically in different
+ways to ensure that a new RS name will never collide with an old one, and use that as the hash
+that shall be appended in the Deployment name to produce the new ReplicaSet name.
```

Comment:
Since we are introducing TemplateGeneration, and that is meant to address problems when comparing if the deployment currently running is the one in Deployment spec, why not just use whatever name (ie. unix time at the moment of creating that RS, similarly to what we do with CronJobs creating Jobs). Although I see value in using TG in the name we are not relying on it, why is it, in that case?